# Limpeza do Dataset




In [ ]:
# Importando Bibliotecas
import pandas as pd
import numpy as np

In [ ]:
# Carregando o Dataset
vendas_cafe1 = pd.read_csv("dirty_cafe_sales.csv")

In [ ]:
vendas_cafe = vendas_cafe1.copy()

In [ ]:
# Dados sujos
vendas_cafe.head(-5)

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11
...,...,...,...,...,...,...,...,...
9990,TXN_1538510,Coffee,5,2.0,10.0,Digital Wallet,NaN,2023-05-22
9991,TXN_3897619,Sandwich,3,4.0,12.0,Cash,Takeaway,2023-02-24
9992,TXN_2739140,Smoothie,4,4.0,16.0,UNKNOWN,In-store,2023-07-05
9993,TXN_4766549,Smoothie,2,4.0,NaN,Cash,NaN,2023-10-20


In [ ]:
# Juntar os nomes das colunas com ifen_
vendas_cafe = vendas_cafe.rename(columns={
    'Transaction ID': 'Transaction_ID',
    'Price Per Unit':'Price_Per_Unit',
    'Total Spent': 'Total_Spent',
    'Payment Method': 'Payment_Method',
    'Transaction Date': 'Transaction_Date'
})

In [ ]:
# Informacoes basicas do Dataset
vendas_cafe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   Transaction_ID    10000 non-null  object
 1   Item              9667 non-null   object
 2   Quantity          9862 non-null   object
 3   Price_Per_Unit    9821 non-null   object
 4   Total_Spent       9827 non-null   object
 5   Payment_Method    7421 non-null   object
 6   Location          6735 non-null   object
 7   Transaction_Date  9841 non-null   object
dtypes: object(8)
memory usage: 625.1+ KB


In [ ]:
#Verificando valores Unicos por colunas e identificando valores incorretos
print(vendas_cafe['Item'].value_counts())
print(vendas_cafe['Quantity'].value_counts())
print(vendas_cafe['Price_Per_Unit'].value_counts())

print(vendas_cafe['Payment_Method'].value_counts())
print(vendas_cafe['Total_Spent'].value_counts())

Item
Juice       1171
Coffee      1165
Salad       1148
Cake        1139
Sandwich    1131
Smoothie    1096
Cookie      1092
Tea         1089
UNKNOWN      344
ERROR        292
Name: count, dtype: int64
Quantity
5          2013
2          1974
4          1863
3          1849
1          1822
UNKNOWN     171
ERROR       170
Name: count, dtype: int64
Price_Per_Unit
3.0        2429
4.0        2331
2.0        1227
5.0        1204
1.0        1143
1.5        1133
ERROR       190
UNKNOWN     164
Name: count, dtype: int64
Payment_Method
Digital Wallet    2291
Credit Card       2273
Cash              2258
ERROR              306
UNKNOWN            293
Name: count, dtype: int64
Total_Spent
6.0        979
12.0       939
3.0        930
4.0        923
20.0       746
15.0       734
8.0        677
10.0       524
2.0        497
9.0        479
5.0        468
16.0       444
25.0       259
7.5        237
1.0        232
4.5        225
1.5        205
UNKNOWN    165
ERROR      164
Name: count, dtype: int64


In [ ]:
# Verificando valores incorretos, porcentegem que representa ao dataset
faltantes_pct = (vendas_cafe.isnull().sum() / len(vendas_cafe)) *100
print(vendas_cafe.isnull().sum())
print('\nPorcentagem de dados incorretos por coluna:')
print(f'{faltantes_pct.sort_values(ascending=False)}')

Transaction_ID         0
Item                 333
Quantity             138
Price_Per_Unit       179
Total_Spent          173
Payment_Method      2579
Location            3265
Transaction_Date     159
dtype: int64

Porcentagem de dados incorretos por coluna:
Location            32.65
Payment_Method      25.79
Item                 3.33
Price_Per_Unit       1.79
Total_Spent          1.73
Transaction_Date     1.59
Quantity             1.38
Transaction_ID       0.00
dtype: float64


In [ ]:
# Substituindo os valores sujos por NaN usando o numpy
#Criado uma lista de valores alem do que foi encontrado
dados_sujo = ['ERROR', 'UNKNOWN', 'None', 'N/A', 'NA', '-', 'null', '']
vendas_cafe = vendas_cafe.replace(dados_sujo, np.nan)

In [ ]:
vendas_cafe.head(-5)

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,NaN,NaN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11
...,...,...,...,...,...,...,...,...
9990,TXN_1538510,Coffee,5,2.0,10.0,Digital Wallet,NaN,2023-05-22
9991,TXN_3897619,Sandwich,3,4.0,12.0,Cash,Takeaway,2023-02-24
9992,TXN_2739140,Smoothie,4,4.0,16.0,NaN,In-store,2023-07-05
9993,TXN_4766549,Smoothie,2,4.0,NaN,Cash,NaN,2023-10-20


In [ ]:
# verificando valores nulos
vendas_cafe.isnull().sum()



,0
Transaction_ID,0
Item,969
Quantity,479
Price_Per_Unit,533
Total_Spent,502
Payment_Method,3178
Location,3961
Transaction_Date,460


In [ ]:
#Verificando valores unicos das colunas categoricas
print(vendas_cafe['Item'].unique())
print(vendas_cafe['Quantity'].unique())
print(vendas_cafe['Payment_Method'].unique())
print(vendas_cafe['Location'].unique())


['Coffee' 'Cake' 'Cookie' 'Salad' 'Smoothie' nan 'Sandwich' 'Juice' 'Tea']
['2' '4' '5' '3' '1' nan]
['Credit Card' 'Cash' nan 'Digital Wallet']
['Takeaway' 'In-store' nan]


In [ ]:
# Mudar os tipos das colunas
vendas_cafe['Item'] = vendas_cafe['Item'].astype('category')
vendas_cafe['Quantity'] = pd.to_numeric(vendas_cafe['Quantity'], errors='coerce').astype('Int64')
vendas_cafe['Price_Per_Unit'] = pd.to_numeric(vendas_cafe['Price_Per_Unit'], errors='coerce').astype('float64')
vendas_cafe['Total_Spent'] = pd.to_numeric(vendas_cafe['Total_Spent'], errors='coerce').astype('float64')
vendas_cafe['Payment_Method'] = vendas_cafe['Payment_Method'].astype('category')
vendas_cafe['Location'] = vendas_cafe['Location'].astype('category')
vendas_cafe['Transaction_Date'] = pd.to_datetime(vendas_cafe['Transaction_Date'], errors='coerce')

In [ ]:
vendas_cafe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction_ID    10000 non-null  object        
 1   Item              9031 non-null   category      
 2   Quantity          9521 non-null   Int64         
 3   Price_Per_Unit    9467 non-null   float64       
 4   Total_Spent       9498 non-null   float64       
 5   Payment_Method    6822 non-null   category      
 6   Location          6039 non-null   category      
 7   Transaction_Date  9540 non-null   datetime64[ns]
dtypes: Int64(1), category(3), datetime64[ns](1), float64(2), object(1)
memory usage: 430.4+ KB


In [ ]:
# Separar Dia da Semana, Mês e Ano
vendas_cafe['Week_Day'] = vendas_cafe['Transaction_Date'].dt.day_name()
vendas_cafe['Day'] = vendas_cafe['Transaction_Date'].dt.day
vendas_cafe['Month'] = vendas_cafe['Transaction_Date'].dt.month
vendas_cafe['Year'] = vendas_cafe['Transaction_Date'].dt.year

In [ ]:
vendas_cafe.head()

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8.0,9.0,2023.0
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16.0,5.0,2023.0
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19,Wednesday,19.0,7.0,2023.0
3,TXN_7034554,Salad,2,5.0,10.0,NaN,NaN,2023-04-27,Thursday,27.0,4.0,2023.0
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11.0,6.0,2023.0


In [ ]:
# Mudar os tipos de coluna Day_week para cateria,Day para int, Month para int e Year para int
vendas_cafe['Transaction_ID'] = vendas_cafe['Transaction_ID'].astype('string')
vendas_cafe['Year'] = vendas_cafe['Year'].astype('Int64')
vendas_cafe['Month'] = vendas_cafe['Month'].astype('Int64')
vendas_cafe['Day'] = vendas_cafe['Day'].astype('Int64')
vendas_cafe['Week_Day'] = vendas_cafe['Week_Day'].astype('category')

In [ ]:
vendas_cafe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction_ID    10000 non-null  string        
 1   Item              9031 non-null   category      
 2   Quantity          9521 non-null   Int64         
 3   Price_Per_Unit    9467 non-null   float64       
 4   Total_Spent       9498 non-null   float64       
 5   Payment_Method    6822 non-null   category      
 6   Location          6039 non-null   category      
 7   Transaction_Date  9540 non-null   datetime64[ns]
 8   Week_Day          9540 non-null   category      
 9   Day               9540 non-null   Int64         
 10  Month             9540 non-null   Int64         
 11  Year              9540 non-null   Int64         
dtypes: Int64(4), category(4), datetime64[ns](1), float64(2), string(1)
memory usage: 704.2 KB


In [ ]:
vendas_cafe.head()

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,NaN,NaN,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023


In [ ]:
# Adicionar nova categoria na Localisação devido porcentagem alta de nulos mais de 30 por cento
# Categoria (Not Specified)
vendas_cafe['Location'] = (
    vendas_cafe['Location'].cat.add_categories('Not Specified').fillna('Not Specified')
)

In [ ]:
vendas_cafe.head()

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,NaN,Not Specified,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023


In [ ]:
# Adicionar nova catgoria no Metodo_Pagamento devido porcentagem alta de nulos mais de 30 porcento
#Categoria (Not Specified)
vendas_cafe['Payment_Method']= (
    vendas_cafe['Payment_Method'].cat.add_categories('Not Specified').fillna('Not Specified')
)


In [ ]:
vendas_cafe.head()

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,Not Specified,Not Specified,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023


In [ ]:
# Tratar volores da coluna 'Quantity' preencher com 1(minimo de venda)
vendas_cafe['Quantity'] = vendas_cafe['Quantity'].fillna(1)

In [ ]:
#Coluna Preco Unitario: preencher com mediana
mediana_por_produto = vendas_cafe.groupby('Item')['Price_Per_Unit'].transform('median')
vendas_cafe['Price_Per_Unit'] = vendas_cafe['Price_Per_Unit'].fillna(mediana_por_produto)

/tmp/ipykernel_6055/2912600762.py:2: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  mediana_por_produto = vendas_cafe.groupby('Item')['Price_Per_Unit'].transform('median')


In [ ]:
vendas_cafe

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,NaN,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,Not Specified,Not Specified,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023
...,...,...,...,...,...,...,...,...,...,...,...,...
9995,TXN_7672686,Coffee,2,2.0,4.0,Not Specified,Not Specified,2023-08-30,Wednesday,30,8,2023
9996,TXN_9659401,NaN,3,NaN,3.0,Digital Wallet,Not Specified,2023-06-02,Friday,2,6,2023
9997,TXN_5255387,Coffee,4,2.0,8.0,Digital Wallet,Not Specified,2023-03-02,Thursday,2,3,2023
9998,TXN_7695629,Cookie,3,1.0,3.0,Digital Wallet,Not Specified,2023-12-02,Saturday,2,12,2023


In [ ]:
# Garantindo o preencimento com mediana casa algum produto tenho todos os precos nulos
vendas_cafe['Price_Per_Unit'] = vendas_cafe['Price_Per_Unit'].fillna(vendas_cafe['Price_Per_Unit'].median())

In [ ]:
# Recalcular Gasto Total
vendas_cafe['Total_Spent'] = vendas_cafe['Quantity'] * vendas_cafe['Price_Per_Unit']

In [ ]:
vendas_cafe

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,4.0,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,Not Specified,Not Specified,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023
...,...,...,...,...,...,...,...,...,...,...,...,...
9995,TXN_7672686,Coffee,2,2.0,4.0,Not Specified,Not Specified,2023-08-30,Wednesday,30,8,2023
9996,TXN_9659401,NaN,3,3.0,9.0,Digital Wallet,Not Specified,2023-06-02,Friday,2,6,2023
9997,TXN_5255387,Coffee,4,2.0,8.0,Digital Wallet,Not Specified,2023-03-02,Thursday,2,3,2023
9998,TXN_7695629,Cookie,3,1.0,3.0,Digital Wallet,Not Specified,2023-12-02,Saturday,2,12,2023


In [ ]:
print(vendas_cafe[[ 'Item','Quantity','Price_Per_Unit',  'Total_Spent','Transaction_Date']].isna().sum())

Item                969
Quantity              0
Price_Per_Unit        0
Total_Spent           0
Transaction_Date    460
dtype: int64


In [ ]:
#Removendo linhas de valores nulos, deixando o dataset limpo, de 10000 linas ficou com 8613
df_limpo = vendas_cafe.dropna(axis=0)

In [ ]:
vendas_cafe1.head()

,Transaction ID,Item,Quantity,Price Per Unit,Total Spent,Payment Method,Location,Transaction Date
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16
2,TXN_4271903,Cookie,4,1.0,ERROR,Credit Card,In-store,2023-07-19
3,TXN_7034554,Salad,2,5.0,10.0,UNKNOWN,UNKNOWN,2023-04-27
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11


In [ ]:
df_limpo.head()

,Transaction_ID,Item,Quantity,Price_Per_Unit,Total_Spent,Payment_Method,Location,Transaction_Date,Week_Day,Day,Month,Year
0,TXN_1961373,Coffee,2,2.0,4.0,Credit Card,Takeaway,2023-09-08,Friday,8,9,2023
1,TXN_4977031,Cake,4,3.0,12.0,Cash,In-store,2023-05-16,Tuesday,16,5,2023
2,TXN_4271903,Cookie,4,1.0,4.0,Credit Card,In-store,2023-07-19,Wednesday,19,7,2023
3,TXN_7034554,Salad,2,5.0,10.0,Not Specified,Not Specified,2023-04-27,Thursday,27,4,2023
4,TXN_3160411,Coffee,2,2.0,4.0,Digital Wallet,In-store,2023-06-11,Sunday,11,6,2023


In [ ]:
df_limpo.info()

<class 'pandas.core.frame.DataFrame'>
Index: 8613 entries, 0 to 9999
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Transaction_ID    8613 non-null   string        
 1   Item              8613 non-null   category      
 2   Quantity          8613 non-null   Int64         
 3   Price_Per_Unit    8613 non-null   float64       
 4   Total_Spent       8613 non-null   Float64       
 5   Payment_Method    8613 non-null   category      
 6   Location          8613 non-null   category      
 7   Transaction_Date  8613 non-null   datetime64[ns]
 8   Week_Day          8613 non-null   category      
 9   Day               8613 non-null   Int64         
 10  Month             8613 non-null   Int64         
 11  Year              8613 non-null   Int64         
dtypes: Float64(1), Int64(4), category(4), datetime64[ns](1), float64(1), string(1)
memory usage: 682.3 KB


# Que foi feito para limpar e normalizar os dados
### No total são 10 mil linhas e 8 colunas
### Foi encontrado dados sujos em 7 colunas em porecentagem (Location            32.65
###Payment Method      25.79
###Item                 3.33
###Price Per Unit       1.79
###Total Spent          1.73
###Transaction Date     1.59
###Quantity             1.38
###Transaction ID       0.00   )
###DAdos sujos encontrados foi ERROR e UNKNOWN

